pkg/controller/deployment/deployment_controller.go

```diff
@@ -424,17 +434,60 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRC, err := dc.getNewRC(deployment)
+// Rolling back to a revision; no-op if the toRevision is deployment's current revision
+func (dc *DeploymentController) rollback(deployment extensions.Deployment, toRevision *int) error {
+	if *toRevision < 0 {
+		dc.eventRecorder.Eventf(&deployment, api.EventTypeWarning, "RollbackDeployment", "Invalid revision to rollback to. Gives up rollback.")
+		// Gives up rollback
+		_, err := dc.cleanUpRollbackTo(&deployment)
+		return err
+	}
+	newRC, allOldRCs, err := dc.getNewRCAndAllOldRCs(deployment)
 	if err != nil {
+		dc.eventRecorder.Eventf(&deployment, api.EventTypeWarning, "RollbackDeployment", "Failed to rollback deployment %s to revision %d: %v", deployment.Name, *toRevision, err)
 		return err
 	}
+	allRCs := append(allOldRCs, newRC)
+	// If rollback revision is 0, rollback to the last revision
+	if *toRevision == 0 {
+		*toRevision = secondMaxRevision(allRCs)
+	}
+	for _, rc := range allRCs {
+		v, err := revision(rc)
+		if err != nil {
+			glog.V(4).Infof("Unable to extract revision from deployment's RC %+v: %v", rc, err)
+			continue
+		}
+		glog.V(4).Infof("Found rc with revision %d: %+v", v, rc)
+		if v == *toRevision {
+			glog.V(4).Infof("Found a match!")
+			// rollback by copying podTemplate.Spec from the rc, and increment revision number by 1
+			// no-op if the the spec matches current deployment's podTemplate.Spec
+			_, err := dc.rollbackToTemplate(&deployment, rc)
+			if err != nil {
+				dc.eventRecorder.Eventf(&deployment, api.EventTypeWarning, "RollbackDeployment", "Failed to rollback deployment %s to revision %d: %v", deployment.Name, *toRevision, err)
+			}
+			return err
+		}
+	}
+	dc.eventRecorder.Eventf(&deployment, api.EventTypeWarning, "RollbackDeployment", "Unable to find the revision to rollback to. Gives up rollback.")
```

Comment:
So we may give up rollback, notify that the rollback didn't happen and in the caller you use another event for notifying that the rollback was successful. I feel that `rollback` needs to return more info to its caller, probably a boolean more indicating if the rollback was done. And the caller should emit all the events.